test/extended/router/externalCertificate.go

```diff
@@ -0,0 +1,682 @@
+			})
+		})
+	})
+
+	g.Context("with valid setup", func() {
+		var (
+			secret       *corev1.Secret
+			routes       []*routev1.Route
+			hosts        []string
+			rootDerBytes []byte
+		)
+
+		g.BeforeEach(func() {
+			const N = 5
+			var routeNames []string
+
+			for i := 0; i < N; i++ {
+				hosts = append(hosts, fmt.Sprintf("host-%d-%s.%s", i, oc.Namespace(), defaultDomain))
+				routeNames = append(routeNames, fmt.Sprintf("route-%d", i))
+			}
+
+			g.By("Creating a secret with crt/key")
+			secret, rootDerBytes, err = createSecret(oc.Namespace(), "my-tls-secret", corev1.SecretTypeTLS, hosts...)
+			o.Expect(err).NotTo(o.HaveOccurred())
+			_, err = oc.KubeClient().CoreV1().Secrets(oc.Namespace()).Create(context.Background(), secret, metav1.CreateOptions{})
+			o.Expect(err).NotTo(o.HaveOccurred())
+
+			g.By("Providing router service account permissions to get,list,watch the secret")
+			_, err = oc.KubeClient().RbacV1().Roles(oc.Namespace()).Create(context.Background(),
+				createRole(oc.Namespace(), "my-tls-secret"), metav1.CreateOptions{})
+			o.Expect(err).NotTo(o.HaveOccurred())
+			_, err = oc.KubeClient().RbacV1().RoleBindings(oc.Namespace()).Create(context.Background(),
+				createRoleBinding(oc.Namespace()), metav1.CreateOptions{})
+			o.Expect(err).NotTo(o.HaveOccurred())
+
+			g.By("Creating multiple routes referencing same external certificate")
+			for i := 0; i < N; i++ {
+				route := createRouteWithEC(oc.Namespace(), routeNames[i], secret.Name, helloPodSvc, hosts[i], routev1.TLSTerminationEdge)
+				_, err = oc.RouteClient().RouteV1().Routes(oc.Namespace()).Create(context.Background(), route, metav1.CreateOptions{})
+				o.Expect(err).NotTo(o.HaveOccurred())
+				routes = append(routes, route)
+			}
+		})
+
+		g.Describe("the router should support external certificate", func() {
+			g.It("and routes are reachable", func() {
+				g.By("Sending https request")
+				for i := 0; i < len(routes); i++ {
+					hostName, err := getHostnameForRoute(oc, routes[i].Name)
+					o.Expect(err).NotTo(o.HaveOccurred())
+					resp, err := httpsGetCall(hostName, rootDerBytes)
+					o.Expect(err).NotTo(o.HaveOccurred())
+					o.Expect(resp).Should(o.ContainSubstring("Hello OpenShift"))
+				}
+			})
+
+			g.Context("and the secret is deleted", func() {
+				g.BeforeEach(func() {
+					g.By("Deleting the secret")
+					err = oc.KubeClient().CoreV1().Secrets(oc.Namespace()).Delete(context.Background(), secret.Name, metav1.DeleteOptions{})
+					o.Expect(err).NotTo(o.HaveOccurred())
+				})
+
+				g.It("then routes are not reachable", func() {
+					g.By("Checking the route status")
+					for i := 0; i < len(routes); i++ {
+						checkRouteStatus(oc, routes[i].Name, corev1.ConditionFalse, "ExternalCertificateValidationFailed")
+					}
+				})
+
+				g.Context("and re-created again", func() {
+					g.BeforeEach(func() {
```

Comment:
In this `Context` only one `It` section is added. This `BeforeEach` section will only be invoked once for the single `It` section. Should this be included inside the `It` section itself?

Similar, trend is there throughout the remaining tests. Not adding multiple comments for them.

Reply:
I think that would help make the tests more structured. 

I wanted to make the `It` blocks more focused (and less cluttered) by adding in necessary checks only (making httpCalls, checking route status etc.) and moving the configuration bits to the `BeforeEach` blocks. 

> In this Context only one It section is added. 

As of now this is true, but when the tests will have to grow in future, I think it would help others IMHO.